zink/codegen/src/event.rs

```diff
@@ -1,19 +1,340 @@
-//! Event interface generation
-
 use proc_macro::{Span, TokenStream};
-use quote::quote;
-use syn::{DeriveInput, LitByteStr};
+use quote::{quote, ToTokens};
+use syn::{
+    parse::Parse, parse_quote, spanned::Spanned, Abi, Data, DeriveInput, Error, Fields, LitByteStr,
+    Result, Type, Variant,
+};
+
+/// Custom error type for better error handling
+#[derive(Debug)]
+enum EventError {
+    NotEnum(Span),
+    UnsupportedType(Span, String),
+    TooManyFields(Span),
+}
 
-/// Expand the event interface
+impl EventError {
+    fn to_compile_error(&self) -> TokenStream {
+        let error_msg = match self {
+            Self::NotEnum(span) => {
+                Error::new((*span).into(), "Event can only be derived for enums")
+            }
+            Self::UnsupportedType(span, ty) => {
+                Error::new((*span).into(), format!("Unsupported type: {}", ty))
+            }
+            Self::TooManyFields(span) => {
+                Error::new((*span).into(), "Too many fields for event logging")
+            }
+        };
+        error_msg.to_compile_error().into()
+    }
+}
+
+/// Expand the event interface with better error handling
 pub fn parse(item: DeriveInput) -> TokenStream {
-    let name = LitByteStr::new(item.ident.to_string().as_bytes(), Span::call_site().into());
-    let ident = item.ident;
+    parse_impl(item)
+}
+
+fn parse_impl(input: DeriveInput) -> TokenStream {
+    let name = &input.ident;
+    let name_bytes = LitByteStr::new(name.to_string().as_bytes(), Span::call_site().into());
+
+    // Ensure we are working with an enum
+    let event_enum = match &input.data {
+        Data::Enum(data_enum) => data_enum,
+        _ => return EventError::NotEnum(proc_macro::Span::call_site()).to_compile_error(),
+    };
+
+    // Generate variant implementations with validation
+    let variant_implementations = match event_enum
+        .variants
+        .iter()
+        .map(|variant| generate_variant_implementation(name, variant))
+        .collect::<Result<Vec<_>>>()
+    {
+        Ok(impls) => impls,
+        Err(e) => return e.to_compile_error().into(),
+    };
+
+    // Generate ABI signature with validation
+    let abi_signature = match generate_abi_signature(name, &event_enum.variants) {
+        Ok(sig) => sig,
+        Err(e) => return e.to_compile_error().into(),
+    };
 
     let expanded = quote! {
-        impl zink::Event for #ident {
-            const NAME: &'static [u8] = #name;
+        impl Event for #name {
+            const NAME: &'static [u8] = #name_bytes;
+
+
+            fn log0(&self) {
+                match self {
+                    #(#variant_implementations)*
+                }
+            }
```

Comment:
we don't need the `Event` trait anymore

```suggestion
        impl #name {
           fn [< #variant0:snake_case >]() {
               unsafe { zink::ffi::log() }
           }
           
           // ...
        }
```

Reply:
impl #name {
            pub fn log0(&self) {
                match self {
                    #(#variant_implementations)*
                }
            }

            pub const fn name() -> &'static [u8] {
                #name_bytes
            }
        }
        
        @clearloop something like this

Reply:
@clearloop something like this :)

Reply:
@clearloop something like this :)

Reply:
https://github.com/zink-lang/zink/pull/298/commits/a8a3eeca309de405e427ddfb1284ed2f62709e7d